flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingSnapshotManagerBase.java

```diff
@@ -0,0 +1,285 @@
+     *
+     * @param fileSystem The filesystem to write to.
+     * @param checkpointBaseDir The base directory for checkpoints.
+     * @param sharedStateDir The directory for shared checkpoint data.
+     * @param taskOwnedStateDir The name of the directory for state not owned/released by the
+     *     master, but by the TaskManagers.
+     * @throws IOException Thrown, if the working directory cannot be created.
+     */
+    private void initFileSystem(
+            FileSystem fileSystem,
+            Path checkpointBaseDir,
+            Path sharedStateDir,
+            Path taskOwnedStateDir)
+            throws IOException {
+        if (fileSystemInitiated) {
+            Preconditions.checkArgument(
+                    checkpointBaseDir.equals(this.checkpointDir),
+                    "The checkpoint base dir is not deterministic across subtasks.");
+            Preconditions.checkArgument(
+                    sharedStateDir.equals(this.sharedStateDir),
+                    "The shared checkpoint dir is not deterministic across subtasks.");
+            Preconditions.checkArgument(
+                    taskOwnedStateDir.equals(this.taskOwnedStateDir),
+                    "The task-owned checkpoint dir is not deterministic across subtasks.");
+            return;
+        }
+        this.fs = fileSystem;
+        this.checkpointDir = checkpointBaseDir;
+        this.sharedStateDir = sharedStateDir;
+        this.taskOwnedStateDir = taskOwnedStateDir;
+        this.fileSystemInitiated = true;
+        this.syncAfterClosingLogicalFile = shouldSyncAfterClosingLogicalFile(checkpointBaseDir);
+        // Initialize the managed exclusive path using id as the child path name.
+        // Currently, we use the task-owned directory to place the merged private state. According
+        // to the FLIP-306, we later consider move these files to the new introduced
+        // task-manager-owned directory.
+        Path managedExclusivePath = new Path(taskOwnedStateDir, id);
+        createManagedDirectory(managedExclusivePath);
+        this.managedExclusiveStateDir = managedExclusivePath;
+    }
+
+    @Override
+    public Path getManagedDir(SubtaskKey subtaskKey, CheckpointedStateScope scope) {
+        if (scope.equals(CheckpointedStateScope.SHARED)) {
+            return managedSharedStateDir.get(subtaskKey);
+        } else {
+            return managedExclusiveStateDir;
+        }
+    }
+
+    static boolean shouldSyncAfterClosingLogicalFile(Path checkpointDir) {
+        // Currently, we do file sync regardless of the file system.
+        return true;
+    }
+
+    // ------------------------------------------------------------------------
+    //  utilities
+    // ------------------------------------------------------------------------
+
+    private void createManagedDirectory(Path managedPath) {
+        try {
+            if (!fs.exists(managedPath)) {
```

Comment:
Here `fs.exists()` and `fs.getFileStatus()` is actually the same method at the underly layer, we could merge them and reduce RPC requests (such as to HDFS), And `isDir` is one deprecated interface, we should switch to `isDirectory`. 
Would something like this work? (NOTE: I have not tested this at all)
```
		boolean isPathExist = false;
		boolean isDirectory = false;
		try {
			FileStatus status = fs.getFileStaus(managedPath);
			isPathExist = true;
			if (status.isDirectory()) {
				isDirectory = true;
			}
		} catch (FileNotFoundException e) {
			isPathExist = false;
		}
		if (!isPathExist) {
			fs.mkdirs(managedPath);
		} else if (!isDirectory) {
			throw new FlinkRuntimeException(
				"The managed path "
					+ managedPath
					+ " for file-merging is occupied by another file. Cannot create directory.");
		}
```

Reply:
You are right that `fs.exists()` is actually calling the `fs.getFileStatus()`. So I changed the logic here. However, we are using the Flink's abstraction of file system (package `org.apache.flink.core.fs`), where `isDir` is the only interface to check directory in `org.apache.flink.core.fs.FileStatus`.